File: pixel_refine_desktop/enhance_stack/core/algorithm/denoising/spatial_core/similarity_taichi/block_matching.py
"""
Block Matching - Taichi GPU
===========================
Mathematical functions for similarity calculation (1:1 parity with C++ block_matching.cpp).
"""
import taichi as ti
import taichi.math as tm

@ti.func
def fast_tanh(x: float) -> float:
    """Fast Tanh approximation matching C++ Padé approximation."""
    res = 0.0
    if x > 3.0:
        res = 1.0
    elif x < -3.0:
        res = -1.0
    else:
        x2 = x * x
        res = x * (27.0 + x2) / (27.0 + 9.0 * x2)
    return res

@ti.func
def calculate_match_confidence(
    mad_score: float,
    noise_sigma: float,
    motion_sensitivity: float,
    noise_offset_factor: float
) -> float:
    """Calculates match confidence using noise-sensitive exponential decay."""
    excess_mad = ti.max(0.0, mad_score - noise_offset_factor * noise_sigma)
    return ti.exp(-excess_mad * motion_sensitivity)

@ti.func
def calculate_hybrid_gradient_optimized(
    current_img: ti.template(),
    reference_img: ti.template(),
    r: int,
    c: int,
    curr_h: int,
    curr_w: int,
    h: int,
    w: int,
    noise_level: float,
    grad_weight_factor: float,
    stab_epsilon: float
) -> float:
    """
    Calculates the hybrid gradient similarity score between current and reference blocks.
    Strict 1:1 parity with C++ Internal::calculate_hybrid_gradient_optimized.
    """
    weighted_sum = 0.0
    total_weight = 0.0
    
    grad_sensitivity = 202.5
    structure_min_threshold_sq = 150.0

    # 1-pixel border skip to prevent out of bounds and match C++
    for y in range(1, curr_h - 1):
        for x in range(1, curr_w - 1):
            img_y = r + y
            img_x = c + x
            
            p1_val = current_img[img_y, img_x]
            p2_val = reference_img[img_y, img_x]
            pixel_diff = ti.abs(p1_val - p2_val)
            
            adaptive_diff_threshold = ti.max(0.005, noise_level * 0.2)
            
            # --- Diagonal Averaging Gradients (3 rows) ---
            # Current Image gradients
            gx1_center = current_img[img_y, img_x + 1] - current_img[img_y, img_x - 1]
            gx1_top = current_img[img_y - 1, img_x + 1] - current_img[img_y - 1, img_x - 1]
            gx1_bottom = current_img[img_y + 1, img_x + 1] - current_img[img_y + 1, img_x - 1]
            gx1 = (gx1_center + gx1_top + gx1_bottom) * 0.333
            
            gy1 = current_img[img_y + 1, img_x] - current_img[img_y - 1, img_x]
            
            # Reference Image gradients
            gx2_center = reference_img[img_y, img_x + 1] - reference_img[img_y, img_x - 1]
            gx2_top = reference_img[img_y - 1, img_x + 1] - reference_img[img_y - 1, img_x - 1]
            gx2_bottom = reference_img[img_y + 1, img_x + 1] - reference_img[img_y + 1, img_x - 1]
            gx2 = (gx2_center + gx2_top + gx2_bottom) * 0.333
            
            gy2 = reference_img[img_y + 1, img_x] - reference_img[img_y - 1, img_x]
            
            mag1_sq = gx1 * gx1 + gy1 * gy1
            mag2_sq = gx2 * gx2 + gy2 * gy2
            min_mag_sq = ti.min(mag1_sq, mag2_sq)
            
            # --- continuous noise weight ---
            noise_weight = 1.0
            if noise_level > stab_epsilon:
                if min_mag_sq < structure_min_threshold_sq:
                    # Flat area
                    local_thr = adaptive_diff_threshold * 1.5
                    if pixel_diff < local_thr:
                        noise_weight = 0.05 + 0.95 * (pixel_diff / local_thr)
                    else:
                        ratio = (pixel_diff - local_thr) / local_thr
                        if ratio > 1.0:
                            ratio = 1.0
                        noise_weight = 1.0 - 0.2 * ratio
                else:
                    # Edge area
                    if pixel_diff < adaptive_diff_threshold:
                        noise_weight = 1.15 + 0.15 * (1.0 - pixel_diff / adaptive_diff_threshold)
                    else:
                        ratio = pixel_diff / (adaptive_diff_threshold * 4.0)
                        if ratio > 1.0:
                            ratio = 1.0
                        noise_weight = 0.3 + 0.4 * (1.0 - ratio)
            
            # --- structure weight ---
            structure_weight = 1.0
            if min_mag_sq > stab_epsilon and mag1_sq > stab_epsilon and mag2_sq > stab_epsilon:
                dot = gx1 * gx2 + gy1 * gy2
                cos_sim = dot / ti.sqrt(mag1_sq * mag2_sq)
                score = ti.max(0.0, cos_sim) * ti.sqrt(min_mag_sq)
                
                structure_weight = 1.0 + grad_weight_factor * fast_tanh(score * grad_sensitivity)
                
            final_weight = structure_weight * noise_weight
            weighted_sum += pixel_diff * final_weight
            total_weight += final_weight

    # Fallback to L1 mean if total weight is very low
    res_val = 0.0
    if total_weight < 1e-4:
        l1_sum = 0.0
        for y in range(curr_h):
            for x in range(curr_w):
                l1_sum += ti.abs(current_img[r + y, c + x] - reference_img[r + y, c + x])
        res_val = l1_sum / float(curr_h * curr_w)
    else:
        res_val = weighted_sum / total_weight

    return res_val

File: pixel_refine_desktop/enhance_stack/core/algorithm/denoising/spatial_core/archive_legacy/similarity_cpp/block_matching.cpp
#include "block_matching.hpp"
#include <algorithm>
#include <cmath>
#include <limits>
#include <opencv2/imgproc.hpp>
#include <vector>

// =========================================================================
// === FAST MATH APPROXIMATIONS (CRITICAL FOR EDGE DEVICES) ===
// =========================================================================

// Aproksimasi Exp cepat (Schraudolph's method like)
// Error < 1-2% tapi 5x-10x lebih cepat dari std::exp
inline float fast_exp(float x) {
  // Clamp untuk mencegah overflow/underflow
  if (x < -80.0f)
    return 0.0f;
  x = 1.0f + x / 256.0f;
  x *= x;
  x *= x;
  x *= x;
  x *= x;
  x *= x;
  x *= x;
  x *= x;
  x *= x;
  return x;
}

// Aproksimasi Tanh cepat menggunakan Pade approximation sederhana atau clamping
// rational
inline float fast_tanh(float x) {
  if (x > 3.0f)
    return 1.0f;
  if (x < -3.0f)
    return -1.0f;
  float x2 = x * x;
  return x * (27.0f + x2) / (27.0f + 9.0f * x2);
}

// =========================================================================
// === BUFFER MANAGEMENT ===
// =========================================================================

// Struktur untuk menampung buffer FFT agar tidak alokasi ulang
struct FFTContext {
  cv::Mat padded1, padded2;
  cv::Mat fft1, fft2;
  cv::Mat diff_dft;
  cv::Mat planes[2];
  cv::Mat mag_sq_diff;
  cv::Size last_size;

  FFTContext() : last_size(0, 0) {}
};

// Thread-local storage agar aman saat parallel processing (OMP)
static thread_local FFTContext fft_ctx;

namespace MotionMatching {
namespace Internal {
static float calculate_plain_mad_32f(const cv::Mat &block1_gray,
                                     const cv::Mat &block2_gray) {
  // === FAST PLAIN MAD (L1 MEAN) ===
  // Replaces Median calculation (std::nth_element) with Mean L1.
  // Why: Sorting is O(N log N) or O(N) average, L1 Mean is O(N) and fully
  // vectorizable. Impact: 2x-3x faster for small blocks.

  // Note: cv::norm(..., NORM_L1) is already AVX optimized by OpenCV.
  if (block1_gray.empty())
    return 0.0f;

  double dist = cv::norm(block1_gray, block2_gray, cv::NORM_L1);
  return static_cast<float>(dist) /
         (float)(block1_gray.rows * block1_gray.cols);
}

static float calculate_fft_32f(const cv::Mat &block1_gray,
                               const cv::Mat &block2_gray, float noise_sigma) {
  // --- OPTIMASI BUFFER FFT ---
  int opt_rows = cv::getOptimalDFTSize(block1_gray.rows);
  int opt_cols = cv::getOptimalDFTSize(block1_gray.cols);

  // Resize buffer hanya jika ukuran berubah
  if (fft_ctx.last_size != cv::Size(opt_cols, opt_rows)) {
    fft_ctx.last_size = cv::Size(opt_cols, opt_rows);
    fft_ctx.padded1.create(opt_rows, opt_cols, CV_32F);
    fft_ctx.padded2.create(opt_rows, opt_cols, CV_32F);
    fft_ctx.fft1.create(opt_rows, opt_cols,
                        CV_32FC1); // Output DFT Real->Complex itu packed
    fft_ctx.fft2.create(opt_rows, opt_cols, CV_32FC1);
  }

  // Zero padding + Copy (Manual optimized copyMakeBorder untuk kasus umum)
  fft_ctx.padded1.setTo(0);
  fft_ctx.padded2.setTo(0);

  cv::Mat roi1 =
      fft_ctx.padded1(cv::Rect(0, 0, block1_gray.cols, block1_gray.rows));
  cv::Mat roi2 =
      fft_ctx.padded2(cv::Rect(0, 0, block2_gray.cols, block2_gray.rows));
  block1_gray.copyTo(roi1);
  block2_gray.copyTo(roi2);

  // Forward DFT
  // Menggunakan DFT_COMPLEX_OUTPUT
  cv::dft(fft_ctx.padded1, fft_ctx.fft1, cv::DFT_COMPLEX_OUTPUT);
  cv::dft(fft_ctx.padded2, fft_ctx.fft2, cv::DFT_COMPLEX_OUTPUT);

  // Compute Difference in Frequency Domain
  cv::subtract(fft_ctx.fft2, fft_ctx.fft1, fft_ctx.diff_dft);
  cv::split(fft_ctx.diff_dft, fft_ctx.planes);

  // Mag Squared: Re^2 + Im^2
  // cv::multiply + cv::add diganti loop manual SIMD friendly
  int n_elements = fft_ctx.planes[0].total();
  if (fft_ctx.mag_sq_diff.total() != n_elements)
    fft_ctx.mag_sq_diff.create(opt_rows, opt_cols, CV_32F);

  const float *re = fft_ctx.planes[0].ptr<float>();
  const float *im = fft_ctx.planes[1].ptr<float>();
  float *dst = fft_ctx.mag_sq_diff.ptr<float>();

#pragma omp simd
  for (int i = 0; i < n_elements; ++i) {
    dst[i] = re[i] * re[i] + im[i] * im[i];
  }

  // Zero out DC
  if (fft_ctx.mag_sq_diff.rows > 0)
    fft_ctx.mag_sq_diff.at<float>(0, 0) = 0.0f;

  // --- LOGIKA WEIGHTING ---
  // (Logika matematika Anda dipertahankan, hanya optimasi akses memori)

  const int meaningful_rows = std::min(opt_rows / 2, block1_gray.rows * 2);
  const int meaningful_cols = std::min(opt_cols / 2, block1_gray.cols * 2);

  const float optimal_elements = static_cast<float>(opt_rows * opt_cols);
  const float noise_sigma_sq = noise_sigma * noise_sigma;

  const float normalized_noise = std::min(1.0f, noise_sigma / 0.13f);
  const float dynamic_noise_floor_factor = 6.0f - (3.0f * normalized_noise);
  const float noise_threshold_sq = std::max(
      1e-6f, noise_sigma_sq * optimal_elements * dynamic_noise_floor_factor);

  const float linear_decay_strength = 1.8f;
  const float row_decay_inv =
      (meaningful_rows > 0) ? (1.0f / meaningful_rows) : 0.0f;
  const float col_decay_inv =
      (meaningful_cols > 0) ? (1.0f / meaningful_cols) : 0.0f;

  float weighted_sum = 0.0f;
  float total_weight = 0.0f;

  // Akses pointer langsung ke mag_sq_diff
  for (int y = 0; y < meaningful_rows; ++y) {
    const float *diff_ptr = fft_ctx.mag_sq_diff.ptr<float>(y);
    float radial_weight =
        std::max(0.0f, 1.0f - (linear_decay_strength * y * row_decay_inv));

    if (radial_weight <= 0.0001f)
      continue; // Early skip row

#pragma omp simd reduction(+ : weighted_sum, total_weight)
    for (int x = 0; x < meaningful_cols; ++x) {
      float col_weight =
          std::max(0.0f, 1.0f - (linear_decay_strength * x * col_decay_inv));
      float pixel_weight = radial_weight * col_weight;

      float val = diff_ptr[x];
      float final_diff =
          (val >= noise_threshold_sq) ? (val - noise_threshold_sq) : 0.0f;

      weighted_sum += final_diff * pixel_weight;
      total_weight += pixel_weight;
    }
  }

  const float total_pixels = block1_gray.rows * block1_gray.cols;
  float fft_mag_sq_score =
      (total_weight > 0) ? (weighted_sum / total_weight) : 0.0f;
  if (total_pixels > 0)
    fft_mag_sq_score /= total_pixels;

  return std::sqrt(fft_mag_sq_score);
}

// --- OPTIMASI HYBRID GRADIENT ---
// Menghapus cv::Scharr dan cv::Laplacian, diganti dengan perhitungan on-the-fly
static float calculate_hybrid_gradient_optimized(const cv::Mat &block1,
                                                 const cv::Mat &block2,
                                                 float noise_level,
                                                 float gradient_weight_factor,
                                                 float stab_epsilon) {
  const int rows = block1.rows;
  const int cols = block1.cols;

  float weighted_sum = 0.0f, total_weight = 0.0f;

  const float grad_sensitivity = 202.5f; // BOOSTED: 150 * 1.35 = 202.5
  // Structure weight optimization: Use fast tanh approximation
  const float structure_thresh_sq = 150.0f;

  // --- OPTIMIZED NOISE WEIGHT PARAMETERS ---
  // Pre-calculate constants for the continuous weighting formula
  // Formula: W = 1.0 / (1.0 + alpha * (diff / tolerance)^2)
  // This replaces the complex branching logic.
  const float tolerance_base = std::max(1e-4f, noise_level);
  // Adaptive scaling: Higher noise = higher tolerance, but slightly clamped
  const float tolerance_scale = tolerance_base * 2.5f;
  const float inv_tolerance_sq = 1.0f / (tolerance_scale * tolerance_scale);

  // Access raw pointers for the whole block range we care about
  // Skipping 1-pixel border to avoid boundary checks
  for (int y = 1; y < rows - 1; ++y) {
    const float *p1 = block1.ptr<float>(y);
    const float *p1_prev = block1.ptr<float>(y - 1);
    const float *p1_next = block1.ptr<float>(y + 1);

    const float *p2 = block2.ptr<float>(y);
    const float *p2_prev = block2.ptr<float>(y - 1);
    const float *p2_next = block2.ptr<float>(y + 1);

// SIMD Friendly Loop: No IF-ELSE branching inside!
#pragma omp simd reduction(+ : weighted_sum, total_weight)
    for (int x = 1; x < cols - 1; ++x) {
      float pixel_diff = std::abs(p1[x] - p2[x]);

      const float adaptive_noise_threshold =
          std::max(0.01f, noise_level * 0.4f);
      const float adaptive_diff_threshold =
          std::max(0.005f, noise_level * 0.2f);
      const float structure_min_threshold_sq = 150.0f;

      // === UPGRADED GRADIENT: DIAGONAL AVERAGING (Option A) ===
      // Averages gradients across 3 rows for ~75% Sobel accuracy
      // Cost: 6 adds + 2 multiplies per direction (vs 2 ops current)

      // Horizontal gradients (3 rows)
      float gx1_center = p1[x + 1] - p1[x - 1];
      float gx1_top = p1_prev[x + 1] - p1_prev[x - 1];
      float gx1_bottom = p1_next[x + 1] - p1_next[x - 1];
      float gx1 = (gx1_center + gx1_top + gx1_bottom) * 0.333f;

      float gx2_center = p2[x + 1] - p2[x - 1];
      float gx2_top = p2_prev[x + 1] - p2_prev[x - 1];
      float gx2_bottom = p2_next[x + 1] - p2_next[x - 1];
      float gx2 = (gx2_center + gx2_top + gx2_bottom) * 0.333f;

      // Vertical gradients (already use 3 columns implicitly via prev/next)
      float gy1 = p1_next[x] - p1_prev[x];
      float gy2 = p2_next[x] - p2_prev[x];

      float mag1_sq = gx1 * gx1 + gy1 * gy1;
      float mag2_sq = gx2 * gx2 + gy2 * gy2;
      float min_mag_sq = (mag1_sq < mag2_sq) ? mag1_sq : mag2_sq;

      // --- RESTORED: ORIGINAL BRANCHING NOISE WEIGHT ---
      float noise_weight = 1.0f;
      if (noise_level > stab_epsilon) {
        if (min_mag_sq < structure_min_threshold_sq) {
          // Flat area
          float local_thr = adaptive_diff_threshold * 1.5f;
          if (pixel_diff < local_thr) {
            noise_weight = 0.05f + 0.95f * (pixel_diff / local_thr);
          } else {
            float ratio = (pixel_diff - local_thr) / local_thr;
            if (ratio > 1.0f)
              ratio = 1.0f;
            noise_weight = 1.0f - 0.2f * ratio;
          }
        } else {
          // Edge area
          if (pixel_diff < adaptive_diff_threshold) {
            noise_weight =
                1.15f + 0.15f * (1.0f - pixel_diff / adaptive_diff_threshold);
          } else {
            float ratio = pixel_diff / (adaptive_diff_threshold * 4.0f);
            if (ratio > 1.0f)
              ratio = 1.0f;
            noise_weight = 0.3f + 0.4f * (1.0f - ratio);
          }
        }
      }

      // 2. STRUCTURE WEIGHT
      float structure_weight = 1.0f;
      if (min_mag_sq > stab_epsilon && mag1_sq > stab_epsilon &&
          mag2_sq > stab_epsilon) {
        float dot = gx1 * gx2 + gy1 * gy2;
        float cos_sim = dot / std::sqrt(mag1_sq * mag2_sq);
        float score = (cos_sim > 0.0f ? cos_sim : 0.0f) * std::sqrt(min_mag_sq);

        structure_weight =
            1.0f + gradient_weight_factor * fast_tanh(score * grad_sensitivity);
      }

      // REMOVED: Texture Weight (Laplacian)

      float final_weight = structure_weight * noise_weight;

      weighted_sum += pixel_diff * final_weight;
      total_weight += final_weight;
    }
  }

  // Fallback for very low weight
  if (total_weight < 1e-4f) {
    // Return simple L1 mean
    return cv::norm(block1, block2, cv::NORM_L1) / (float)(rows * cols);
  }

  return weighted_sum / total_weight;
}

} // namespace Internal

// =========================================================================
// === PUBLIC INTERFACE IMPLEMENTATION ===
// =========================================================================

TileMatchResult
calculate_tile_mad(const cv::Mat &current_tile_gray,
                   const cv::Mat &reference_tile_gray, float global_noise_sigma,
                   float gradient_weight_factor, float stability_epsilon,
                   MBMBuffers &buffers) // Parameter buffers tidak lagi krusial
                                        // tapi disimpan utk kompatibilitas
{
  TileMatchResult result;

  // Check validity
  if (current_tile_gray.empty() || reference_tile_gray.empty()) {
    result.success = false;
    return result;
  }

  const int bh = current_tile_gray.rows;
  const int bw = current_tile_gray.cols;

  // ALWAYS use hybrid gradient for best quality
  // Removed plain MAD fallback as per user request
  result.mad_score = Internal::calculate_hybrid_gradient_optimized(
      current_tile_gray, reference_tile_gray, global_noise_sigma,
      gradient_weight_factor, stability_epsilon);

  result.success = true;
  return result;
}

TileMatchResult calculate_tile_fft(const cv::Mat &current_tile_gray,
                                   const cv::Mat &reference_tile_gray,
                                   float global_noise_sigma) {
  TileMatchResult result;

  if (current_tile_gray.empty()) {
    result.success = false;
    return result;
  }

  const int bh = current_tile_gray.rows;

  // FFT overhead terlalu besar untuk blok < 8x8.
  if (bh < 8) {
    result.mad_score = Internal::calculate_plain_mad_32f(current_tile_gray,
                                                         reference_tile_gray);
  } else {
    result.mad_score = Internal::calculate_fft_32f(
        current_tile_gray, reference_tile_gray, global_noise_sigma);
  }

  result.success = true;
  return result;
}

} // namespace MotionMatching